### DISCUSSION TITLE
Why is the musicxml converted to json through alphatab, and the parsed PlaybackInformation.Program  is all 1

### DISCUSSION THREAD

**Discussion by phpsxg** [CoderLine/alphaTab#665] [Q&A] [answered]
I converted musicxml to json through alphatab, but found that the parsed PlaybackInformation.Program is all 1. How to correctly parse out the instrument, Looking forward to your reply，Thanks

English is not my native language; please excuse typing errors.

通过alphatab，将musicxml文件转化为json,发现不同乐器品类转化json后，PlaybackInformation.Program 都为1，是程序哪里有问题吗，如何正确解析出乐器品类呢，谢谢

```
MusicXmlUtils musicxml = new MusicXmlUtils();
string json = JsonConvert.SerializeObject(musicxml.export_json(file_name), Formatting.Indented, new JsonSerializerSettings { ReferenceLoopHandling = ReferenceLoopHandling.Ignore });
                            
```

**Comment by Danielku15** [comment, marked as answer]
It depends strongly on your MusicXML files. Depending on the software you are using for MusicXML, it might not write correctly the midi programs. AlphaTab is loading the data from the [midi-program](https://www.w3.org/2021/06/musicxml40/musicxml-reference/elements/midi-program/) tag in MusicXML. If you ensure this data is filled, alphaTab will respect this information. 

While MusicXML is advertised as a standard for exchanging music sheets, it seems that almost every software has some custom proprietary parts which break the exchange.

**Reply by phpsxg** [reply]
I have another problem. When converting xml to musicxml through alphatab, Chords.String is empty in the parsed json.

xml
>  <harmony>
    <root>
     <root-step>F</root-step>
     <root-alter>0</root-alter>
    </root>
    <kind>minor</kind>
    <frame>
     <frame-strings>6</frame-strings>
     <frame-frets>5</frame-frets>
     <first-fret>1</first-fret>
     <frame-note>
      <string>6</string>
      <fret>1</fret>
     </frame-note>
     <frame-note>
      <string>5</string>
      <fret>3</fret>
     </frame-note>
     <frame-note>
      <string>4</string>
      <fret>3</fret>
     </frame-note>
     <frame-note>
      <string>3</string>
      <fret>1</fret>
     </frame-note>
     <frame-note>
      <string>2</string>
      <fret>1</fret>
     </frame-note>
     <frame-note>
      <string>1</string>
      <fret>1</fret>
     </frame-note>
    </frame>
   </harmony>

json

>    "Chords": {
            "FA7A18F0-3C2B-CBD8-F692-FC76644F02A2": {
              "Name": "C",
              "FirstFret": 1.0,
              "Strings": [],
              "BarreFrets": [],
              "ShowName": true,
              "ShowDiagram": true,
              "ShowFingering": true
            },
            "6EE3CB6B-AB91-EF77-F629-FDBEE4455F4D": {
              "Name": "F",
              "FirstFret": 1.0,
              "Strings": [],
              "BarreFrets": [],
              "ShowName": true,
              "ShowDiagram": true,
              "ShowFingering": true
            }
          }

Check the ParseHarmony method in the alphatab c# source importer/MusicXmlImporter.cs, and found that only the root and kind are judged, and the frame is not processed. Is it xml not supported?

**Reply by Danielku15** [reply]
Can you provide some full example files so we can have a look at what you're reporting. Best attach them as files and write which errors you see on which files. This way I can runs some tests. 

It is hard to judge without data if there might be a bug in alphaTab or if your input data is malformed.

**Reply by phpsxg** [reply]
> Can you provide some full example files so we can have a look at what you're reporting. Best attach them as files and write which errors you see on which files. This way I can runs some tests.
> 
> It is hard to judge without data if there might be a bug in alphaTab or if your input data is malformed.


The demo xml url is:http://39.105.196.188:8080/demo.xml, when converting demo xml to json through alphatab, Chords.String is empty in the parsed json.

C# code for converting xml to json
```
var musicxml = new MusicXmlImporter();
var settings = new Settings();
musicxml.Init(ByteBuffer.FromBuffer(File.ReadAllBytes(file_name)),settings);
var score = musicxml.ReadScore();
string json = JsonConvert.SerializeObject(score, Formatting.Indented,
    new JsonSerializerSettings { ReferenceLoopHandling = ReferenceLoopHandling.Ignore });
File.WriteAllText(file_name, json);
```

**Reply by Danielku15** [reply]
From the MusicXML I see that you exported the file from Guitar Pro, in this case you might want to consider using Guitar Pro files directly instead of going through MusicXML. alphaTab can read Guitar Pro 3-7 files directly. 

The answer to your question is: alphaTab does not support reading the full chord information from MusicXML. It is a missing feature. 
https://github.com/CoderLine/alphaTab/blob/917f90f21f287c29ae9379d6adf7ba3adb29df66/src/importer/MusicXmlImporter.ts#L464

I added https://github.com/CoderLine/alphaTab/issues/667 as work item to add support but cannot promise to deliver this feature in near future. Contributions are welcome.

**Reply by phpsxg** [reply]
@Danielku15 Convert musicxml to json through alphatab. How to get the numbered musical notation according to the json file? What are the corresponding rules?
According to the two values of note.Tone  and note.Octave, can the corresponding numbered musical notation be obtained? What are the rules?
Thanks!

**Reply by Danielku15** [reply]
I recommend that you read more about how to use the raw data model of alphatab. If you want to access the details of the loaded file, you should not convert it to json but keep the Score instance and use it to read the data you are interested into. 

the JSON converter is primarily meant for internal usage when passing the data between components when serialization is needed. within the same application, just use the classes as available.

The classes of the data model have a lot of helpful properties which allow you to get information like the absolute midi value a note will sound like (respecting all effects and configurations). You might want to browse through the code of these classes as well to understand what properties exist and how they are computed. 


https://www.alphatab.net/docs/reference/score
https://github.com/CoderLine/alphaTab/tree/develop/src%2Fmodel
https://www.alphatab.net/docs/guides/song-details
https://www.alphatab.net/docs/tutorial-net/introduction
